sdmetrics/single_table/privacy/dcr_utils.py

```diff
@@ -1,156 +1,114 @@
 
-    real_data_copy = real_data.copy()
-    synthetic_data_copy = synthetic_data.copy()
-    real_data_copy = _process_data_with_metadata(real_data_copy, metadata, True)
-    synthetic_data_copy = _process_data_with_metadata(synthetic_data_copy, metadata, True)
+    common_cols = set(dataset_copy.columns) & set(reference_copy.columns)
+    cols_to_keep = []
+    ranges = {}
 
-    overlapping_columns = set(real_data_copy.columns) & set(synthetic_data_copy.columns)
-    if not overlapping_columns:
-        raise ValueError('There are no overlapping statistical columns to measure.')
+    for col_name, col_metadata in get_columns_from_metadata(metadata).items():
+        sdtype = col_metadata['sdtype']
+
+        if (
+            sdtype in ['numerical', 'categorical', 'boolean', 'datetime']
+            and col_name in common_cols
+        ):
+            cols_to_keep.append(col_name)
 
-    for col_name, column in get_columns_from_metadata(metadata).items():
-        sdtype = column['sdtype']
-        col_range = None
-        if sdtype == 'numerical' or sdtype == 'datetime':
-            col_range = real_data_copy[col_name].max() - real_data_copy[col_name].min()
-            if isinstance(col_range, pd.Timedelta):
-                col_range = col_range.total_seconds()
+            if sdtype in ['numerical', 'datetime']:
+                col_range = reference_copy[col_name].max() - reference_copy[col_name].min()
+                if isinstance(col_range, pd.Timedelta):
+                    col_range = col_range.total_seconds()
+                ranges[col_name] = col_range
 
-        column_ranges[col_name] = col_range
+    if not cols_to_keep:
+        raise ValueError('There are no overlapping statistical columns to measure.')
 
-    dcr_dist_df = synthetic_data_copy.apply(
-        lambda synth_row: _calculate_dcr_between_row_and_data(
-            synth_row, real_data_copy, column_ranges, metadata
-        ),
-        axis=1,
-    )
+    # perform a full cross join on the data we want to compute
+    dataset_copy = dataset_copy[cols_to_keep]
+    dataset_copy['index'] = [i for i in range(len(dataset_copy))]
+
+    reference_copy = reference_copy[cols_to_keep]
+    reference_copy['index'] = [i for i in range(len(reference_copy))]
+
+    results = []
+    chunk_size = 1000
+
+    for chunk_start in range(0, len(dataset_copy), chunk_size):
+        chunk = dataset_copy.iloc[chunk_start : chunk_start + chunk_size].copy()
+        chunk['index'] = range(chunk_start, chunk_start + len(chunk))
+
+        reference_copy['index'] = range(len(reference_copy))
+        full_dataset = chunk.merge(reference_copy, how='cross', suffixes=('_data', '_ref'))
+
+        for col_name in cols_to_keep:
+            sdtype = metadata['columns'][col_name]['sdtype']
+            if sdtype in ['numerical', 'datetime']:
+                diff = (full_dataset[col_name + '_ref'] - full_dataset[col_name + '_data']).abs()
```

Comment:
Since we're indexing into the datasets a lot, it might be a bit easier to read if we assigned the columns we're working with to `ref_column` and `data_column` and used those variables instead